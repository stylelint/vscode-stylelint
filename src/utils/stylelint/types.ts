import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages/index';
export type Stylelint = typeof stylelint;
export type ConfigurationError = Error & { code: 78 };

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;

	/**
	 * Gets the original warning from the given diagnostic.
	 */
	getWarning?: (diagnostic: LSP.Diagnostic) => stylelint.Warning | null;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}
