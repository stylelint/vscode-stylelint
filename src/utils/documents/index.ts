export * from './create-text-edits';
export * from './get-edit-info';
export * from './get-disable-type';
export * from './get-fixes';
export * from './get-workspace-folder';
