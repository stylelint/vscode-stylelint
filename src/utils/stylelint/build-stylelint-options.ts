import path from 'path';
import pathIsInside from 'path-is-inside';
import { URI } from 'vscode-uri';
import { findPackageRoot } from '../packages/index';
import type stylelint from 'stylelint';
import type { RunnerOptions } from './types';

/**
 * Given a document URI, base options, and extension options, builds a Stylelint
 * options object. Runner options supersede base options.
 */
export async function buildStylelintOptions(
	uri: string,
	workspaceFolder?: string,
	baseOptions: Partial<stylelint.LinterOptions> = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportDescriptionlessDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	}: RunnerOptions = {},
): Promise<Partial<stylelint.LinterOptions>> {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportDescriptionlessDisables:
			reportDescriptionlessDisables ?? baseOptions.reportDescriptionlessDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.stylelintignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.stylelintignore',
			);
		}
	}

	return options;
}
