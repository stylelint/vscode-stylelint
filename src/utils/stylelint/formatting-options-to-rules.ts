import type LSP from 'vscode-languageserver-protocol';

export type FormattingRules = {
	indentation?: [number | string];
	'no-missing-end-of-source-newline'?: true | null;
	'no-eol-whitespace'?: true | null;
};

/**
 * Converts the given formatting options to rules.
 * @param options The formatting options.
 * @returns The rules.
 */
export function formattingOptionsToRules({
	insertSpaces,
	tabSize,
	insertFinalNewline,
	trimTrailingWhitespace,
}: LSP.FormattingOptions): FormattingRules {
	// NOTE: There is no equivalent rule for trimFinalNewlines, so it is not respected.
	// TODO: Create respective rule upstream?

	const rules: FormattingRules = {
		indentation: [insertSpaces ? tabSize : 'tab'],
	};

	if (insertFinalNewline !== undefined) {
		rules['no-missing-end-of-source-newline'] = insertFinalNewline || null;
	}

	if (trimTrailingWhitespace !== undefined) {
		rules['no-eol-whitespace'] = trimTrailingWhitespace || null;
	}

	return rules;
}
