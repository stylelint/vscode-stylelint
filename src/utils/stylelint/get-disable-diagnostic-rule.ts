import type LSP from 'vscode-languageserver-protocol';
import { DisableReportRuleNames } from './types';

/**
 * Gets the rule name to which a disable diagnostic applies. Returns `undefined`
 * if the diagnostic is not a disable diagnostic.
 * @param diagnostic The diagnostic corresponding to the Stylelint warning.
 */
export function getDisableDiagnosticRule(diagnostic: LSP.Diagnostic): string | undefined {
	// eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
	switch (diagnostic.code) {
		case DisableReportRuleNames.Needless:
			return diagnostic.message.match(/^Needless disable for "(.+)"$/)?.[1];

		case DisableReportRuleNames.InvalidScope:
			return diagnostic.message.match(/^Rule "(.+)" isn't enabled$/)?.[1];

		case DisableReportRuleNames.Descriptionless:
			return diagnostic.message.match(/^Disable for "(.+)" is missing a description$/)?.[1];

		case DisableReportRuleNames.Illegal:
			return diagnostic.message.match(/^Rule "(.+)" may not be disabled$/)?.[1];

		default:
			return undefined;
	}
}
