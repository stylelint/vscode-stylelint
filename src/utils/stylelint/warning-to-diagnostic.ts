import { Diagnostic, DiagnosticSeverity, Position, Range } from 'vscode-languageserver-types';
import type stylelint from 'stylelint';

/**
 * Converts a Stylelint warning to an LSP Diagnostic.
 *
 * @example
 * ```js
 * const [result] = await stylelint.lint({
 *   code: 'a { color: red; }',
 *   config: { rules: { 'color-named': 'never' } }
 * });
 *
 * const [warning] = result.warnings;
 * // {
 * //   rule: 'color-named',
 * //   text: 'Unexpected named color "red" (color-named)',
 * //   severity: 'error',
 * //   line: 1,
 * //   column: 12
 * // }
 *
 * const diagnostic = warningToDiagnostic(warning);
 * // {
 * //   message: 'Unexpected named color "red" (color-named)',
 * //   severity: 1,
 * //   source: 'Stylelint',
 * //   range: {
 * //     start: {
 * //       line: 0,
 * //       character: 11
 * //     },
 * //     end: {
 * //       line: 0,
 * //       character: 11
 * //     }
 * //   }
 * // }
 * ```
 * @param warning The warning to convert.
 * @param rules Available Stylelint rules.
 */
export function warningToDiagnostic(
	warning: stylelint.Warning,
	ruleMetadata?: stylelint.LinterResult['ruleMetadata'],
): Diagnostic {
	const start = Position.create(warning.line - 1, warning.column - 1);
	const end =
		typeof warning.endLine === 'number' && typeof warning.endColumn === 'number'
			? Position.create(warning.endLine - 1, warning.endColumn - 1)
			: Position.create(warning.line - 1, warning.column);

	const ruleDocUrl = ruleMetadata?.[warning.rule]?.url;

	const diagnostic = Diagnostic.create(
		Range.create(start, end),
		warning.text,
		DiagnosticSeverity[warning.severity === 'warning' ? 'Warning' : 'Error'],
		warning.rule,
		'Stylelint',
	);

	if (ruleDocUrl) {
		diagnostic.codeDescription = { href: ruleDocUrl };
	}

	return diagnostic;
}
