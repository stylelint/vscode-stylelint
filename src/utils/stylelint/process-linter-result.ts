import * as crypto from 'crypto';
import { warningToDiagnostic } from './warning-to-diagnostic';
import type { LinterResult, Warning } from 'stylelint';
import { type LintDiagnostics, type Stylelint, InvalidOptionError } from './types';
import type LSP from 'vscode-languageserver-protocol';

/**
 * Returns a unique key for a diagnostic.
 * @param diagnostic The diagnostic to get a key for.
 */
function getDiagnosticKey(diagnostic: LSP.Diagnostic): string {
	const range = diagnostic.range;
	let message = '';

	if (diagnostic.message) {
		const hash = crypto.createHash('sha256');

		hash.update(diagnostic.message);
		message = hash.digest('base64');
	}

	return `[${range.start.line},${range.start.character},${range.end.line},${range.end.character}]-${diagnostic.code}-${message}`;
}

/**
 * Processes the results of a Stylelint lint run.
 *
 * If Stylelint reported any warnings, they are converted to Diagnostics and
 * returned. If the lint results contain raw output in the `output` property, it
 * is also returned.
 *
 * Throws an `InvalidOptionError` for any invalid option warnings reported by
 * Stylelint.
 * @param stylelint The Stylelint instance that was used.
 * @param result The results returned by Stylelint.
 */
export function processLinterResult(
	stylelint: Stylelint,
	linterResult: LinterResult,
): LintDiagnostics {
	const { results } = linterResult;

	if (results.length === 0) {
		return { diagnostics: [] };
	}

	const [{ invalidOptionWarnings, warnings, ignored }] = results;

	if (ignored) {
		return { diagnostics: [] };
	}

	if (invalidOptionWarnings.length !== 0) {
		throw new InvalidOptionError(invalidOptionWarnings);
	}

	let { ruleMetadata } = linterResult;

	if (!ruleMetadata) {
		// Create built-in rule metadata for backwards compatibility.
		ruleMetadata = new Proxy(
			{},
			{
				get: (_, key: string) => {
					// eslint-disable-next-line @typescript-eslint/ban-ts-comment
					// @ts-ignore -- (TS7053) `stylelint.rules` has returned `Promise` values since v16.
					// See https://stylelint.io/migration-guide/to-16#changed-nodejs-api-stylelintrules-object
					//
					// eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
					return stylelint.rules?.[key]?.meta;
				},
			},
		);
	}

	const diagnostics: LSP.Diagnostic[] = [];
	const warningsMap = new Map<string, Warning>();

	for (const warning of warnings) {
		const diagnostic = warningToDiagnostic(warning, ruleMetadata);

		diagnostics.push(diagnostic);
		warningsMap.set(getDiagnosticKey(diagnostic), warning);
	}

	const getWarning = (diagnostic: LSP.Diagnostic): Warning | null => {
		const key = getDiagnosticKey(diagnostic);

		return warningsMap.get(key) ?? null;
	};
	const output = ('report' in linterResult && linterResult.report) || linterResult.output;

	return output ? { output, diagnostics, getWarning } : { diagnostics, getWarning };
}
