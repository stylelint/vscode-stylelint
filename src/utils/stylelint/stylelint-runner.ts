import os from 'os';
import { URI } from 'vscode-uri';
import type { Connection } from 'vscode-languageserver';
import type { TextDocument } from 'vscode-languageserver-textdocument';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type winston from 'winston';

import { StylelintResolver } from '../packages/index';
import { getWorkspaceFolder } from '../documents/index';
import { processLinterResult } from './process-linter-result';
import { buildStylelintOptions } from './build-stylelint-options';
import type { LintDiagnostics, RunnerOptions } from './types';

/**
 * Runs Stylelint in VS Code.
 */
export class StylelintRunner {
	/**
	 * The language server connection.
	 */
	#connection: Connection | undefined;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * The Stylelint resolver.
	 */
	#stylelintResolver: StylelintResolver;

	constructor(connection?: Connection, logger?: winston.Logger, resolver?: StylelintResolver) {
		this.#connection = connection;
		this.#logger = logger;
		this.#stylelintResolver = resolver ?? new StylelintResolver(connection, logger);
	}

	/**
	 * Lints the given document using Stylelint. The linting result is then
	 * converted to LSP diagnostics and returned.
	 * @param document
	 * @param linterOptions
	 * @param extensionOptions
	 */
	async lintDocument(
		document: TextDocument,
		linterOptions: stylelint.LinterOptions = {},
		runnerOptions: RunnerOptions = {},
	): Promise<LintDiagnostics> {
		const workspaceFolder =
			this.#connection && (await getWorkspaceFolder(this.#connection, document));

		const result = await this.#stylelintResolver.resolve(runnerOptions, document);

		if (!result) {
			this.#logger?.info('No Stylelint found with which to lint document', {
				uri: document.uri,
				options: runnerOptions,
			});

			return { diagnostics: [] };
		}

		const { stylelint } = result;

		const { fsPath } = URI.parse(document.uri);

		// Workaround for Stylelint treating paths as case-sensitive on Windows
		// If the drive letter is lowercase, we need to convert it to uppercase
		// See https://github.com/stylelint/stylelint/issues/5594
		// TODO: Remove once fixed upstream
		const codeFilename =
			os.platform() === 'win32'
				? fsPath.replace(/^[a-z]:/, (match) => match.toUpperCase())
				: fsPath;

		const options: stylelint.LinterOptions = {
			...(await buildStylelintOptions(document.uri, workspaceFolder, linterOptions, runnerOptions)),
			code: document.getText(),
			formatter: () => '',
		};

		if (codeFilename) {
			options.codeFilename = codeFilename;
		} else if (!linterOptions?.config?.rules) {
			options.config = { rules: {} };
		}

		if (this.#logger?.isDebugEnabled()) {
			this.#logger?.debug('Running Stylelint', {
				options: { ...options, code: '...' },
			});
		}

		try {
			return processLinterResult(stylelint, await stylelint.lint(options));
		} catch (err) {
			if (
				err instanceof Error &&
				(err.message.startsWith('No configuration provided for') ||
					err.message.includes('No rules found within configuration'))
			) {
				// Check only CSS syntax errors without applying any Stylelint rules
				return processLinterResult(
					stylelint,
					await stylelint.lint({ ...options, config: { rules: {} } }),
				);
			}

			throw err;
		}
	}
}
