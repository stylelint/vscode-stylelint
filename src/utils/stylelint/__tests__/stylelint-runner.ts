jest.mock('os');
jest.mock('path');
jest.mock('vscode-uri', () => ({
	URI: {
		parse: jest.fn((str: string) => ({ fsPath: str })),
	},
}));
jest.mock('../../packages/stylelint-resolver');
jest.mock('../../documents');

import os from 'os';
import path from 'path';
import stylelint from 'stylelint';
import type winston from 'winston';
import type { Connection } from 'vscode-languageserver';
import type { TextDocument } from 'vscode-languageserver-textdocument';
import * as LSP from 'vscode-languageserver-protocol';
import { StylelintResolver, ResolverOptions } from '../../packages/index';
import { getWorkspaceFolder } from '../../documents/index';
import { StylelintRunner } from '../stylelint-runner';
import { version as stylelintVersion } from 'stylelint/package.json';
import semver from 'semver';

const mockedOS = os as tests.mocks.OSModule;
const mockedPath = path as tests.mocks.PathModule;
const mockedResolver = StylelintResolver as jest.Mock<StylelintResolver>;
const mockedGetWorkspaceFolder = getWorkspaceFolder as jest.MockedFunction<
	typeof getWorkspaceFolder
>;

type FakeLintFunction = (
	options: stylelint.LinterOptions,
) => Promise<Partial<stylelint.LinterResult>>;

type FakeResolveFunction = (
	serverOptions: ResolverOptions,
	document: TextDocument,
) => Promise<{ stylelint: { lint: FakeLintFunction } }>;

const createMockResolver =
	(lint?: FakeLintFunction, resolve?: FakeResolveFunction): (() => StylelintResolver) =>
	() =>
		({
			resolve: resolve ?? (() => Promise.resolve(lint ? { stylelint: { lint } } : undefined)),
		}) as unknown as StylelintResolver;

const createMockDocument = (code: string, uri = '/path/to/file.css'): TextDocument =>
	({
		getText: () => code,
		uri,
	}) as TextDocument;

const mockConnection = {} as unknown as Connection;

describe('StylelintRunner', () => {
	beforeEach(() => {
		mockedOS.__mockPlatform('linux');
		mockedPath.__mockPlatform('posix');
	});

	test('should return no diagnostics if Stylelint cannot be resolved', async () => {
		mockedResolver.mockImplementation(createMockResolver());

		const results = await new StylelintRunner(mockConnection).lintDocument(createMockDocument(''));

		expect(results).toEqual({ diagnostics: [] });
	});

	// TODO: Remove once fixed upstream
	test('should upper-case drive letters on Windows (Stylelint bug #5594)', async () => {
		expect.assertions(2);

		mockedOS.__mockPlatform('win32');
		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.codeFilename).toBe('C:\\path\\to\\file.css');

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('', 'c:\\path\\to\\file.css'),
		);

		mockedOS.__mockPlatform('linux');
		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.codeFilename).toBe('c:/path/to/file.css');

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('', 'c:/path/to/file.css'),
		);
	});

	test('should call stylelint.lint with the document path and given options', async () => {
		expect.assertions(2);

		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options).toMatchSnapshot();

				expect((options.formatter as stylelint.Formatter)?.([], {} as never)).toBe('');

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.scss'),
			{
				config: {
					customSyntax: 'postcss-scss',
				},
				fix: true,
			},
		);
	});

	test("should pass empty rules if the document's path cannot be determined and rules aren't set", async () => {
		expect.assertions(1);

		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.config).toEqual({
					rules: {},
				});

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(createMockDocument('a {}', ''));
	});

	test("should not change set rules if the document's path cannot be determined", async () => {
		expect.assertions(1);

		mockedResolver.mockImplementation(
			createMockResolver(async (options) => {
				expect(options.config).toEqual({
					rules: { 'no-empty-source': true },
				});

				return { results: [] };
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(createMockDocument('a {}', ''), {
			config: { rules: { 'no-empty-source': true } },
		});
	});

	test('should call the resolver with the server options and the document', async () => {
		expect.assertions(3);

		mockedResolver.mockImplementation(
			createMockResolver(undefined, async (serverOptions, document) => {
				expect(serverOptions).toEqual({
					packageManager: 'pnpm',
				});

				expect(document).toEqual({
					getText: expect.any(Function),
					uri: '/path/to/file.css',
				});

				expect(document.getText()).toBe('a {}');

				return {
					stylelint: { lint: async () => ({ results: [] }) },
				};
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			undefined,
			{ packageManager: 'pnpm' },
		);
	});

	test('with stylelintPath, should call the resolver with the path', async () => {
		expect.assertions(1);

		mockedGetWorkspaceFolder.mockResolvedValueOnce('/workspace');

		mockedResolver.mockImplementation(
			createMockResolver(undefined, async (serverOptions) => {
				expect(serverOptions).toEqual({
					stylelintPath: '/path/to/stylelint',
				});

				return {
					stylelint: { lint: async () => ({ results: [] }) },
				};
			}),
		);

		await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			undefined,
			{ stylelintPath: '/path/to/stylelint' },
		);
	});

	test('should return processed lint results from Stylelint without configured rules', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ stylelint })));

		const results = await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('table {', '/path/to/file.css'),
		);

		expect(results).toMatchSnapshot();
	});

	test('should return processed lint results from Stylelint with configured rules', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ stylelint })));

		const results = await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			{ config: { rules: { 'block-no-empty': true } } },
		);

		expect(results).toMatchSnapshot();
	});

	test('should throw errors thrown by Stylelint', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ stylelint })));

		await expect(
			new StylelintRunner(mockConnection).lintDocument(
				createMockDocument('a {}', '/path/to/file.css'),
				{
					config: { rules: {} },
					files: ['/path/to/file.css'],
				},
			),
		).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should log if a logger is provided', async () => {
		expect.assertions(2);

		const mockLogger = {
			info: jest.fn(),
			warn: jest.fn(),
			error: jest.fn(),
			debug: jest.fn(),
			isDebugEnabled: jest.fn(() => true),
		} as unknown as winston.Logger;

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ stylelint })));

		await new StylelintRunner(mockConnection, mockLogger).lintDocument(
			createMockDocument('a {}', '/path/to/file.css'),
			{ config: { rules: { 'block-no-empty': true } } },
		);

		expect(mockLogger.debug).toHaveBeenCalledTimes(1);
		expect(mockLogger.debug).toHaveBeenCalledWith(
			expect.stringMatching(/^Running Stylelint/),
			expect.any(Object),
		);
	});

	if (semver.satisfies(stylelintVersion, '>=16.15')) {
		it('should be possible to get the original warning via getWarning', async () => {
			expect.assertions(1);

			mockedPath.__mockPlatform();

			mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ stylelint })));

			const results = await new StylelintRunner(mockConnection).lintDocument(
				createMockDocument('a {color:#fff}', '/path/to/file.css'),
				{ config: { rules: { 'color-hex-length': 'long' } } },
			);

			expect(results.getWarning?.(results.diagnostics[0])).toEqual({
				column: 10,
				endColumn: 14,
				endLine: 1,
				fix: {
					range: [12, 13],
					text: 'ffff',
				},
				line: 1,
				rule: 'color-hex-length',
				severity: 'error',
				text: `Expected "#fff" to be "#ffffff" (color-hex-length)`,
				url: undefined,
			});
		});
	}

	it('should not get the warning with getWarning and a non-existent diagnostic', async () => {
		expect.assertions(1);

		mockedPath.__mockPlatform();

		mockedResolver.mockImplementation(createMockResolver(undefined, async () => ({ stylelint })));

		const results = await new StylelintRunner(mockConnection).lintDocument(
			createMockDocument('a {color:#fff}', '/path/to/file.css'),
			{ config: { rules: { 'color-hex-length': 'long' } } },
		);

		expect(
			results.getWarning?.({
				message: 'Message for rule 1',
				source: 'Stylelint',
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
				code: 'rule 1',
				severity: LSP.DiagnosticSeverity.Error,
				codeDescription: {
					href: 'https://stylelint.io/user-guide/rules/rule',
				},
			}),
		).toMatchSnapshot();
	});
});
