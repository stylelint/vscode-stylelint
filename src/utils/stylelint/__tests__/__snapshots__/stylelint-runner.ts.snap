// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StylelintRunner should call stylelint.lint with the document path and given options 1`] = `
{
  "code": "a {}",
  "codeFilename": "/path/to/file.scss",
  "config": {
    "customSyntax": "postcss-scss",
  },
  "configBasedir": undefined,
  "configFile": undefined,
  "customSyntax": undefined,
  "fix": true,
  "formatter": [Function],
  "ignoreDisables": undefined,
  "ignorePath": "/.stylelintignore",
  "reportDescriptionlessDisables": undefined,
  "reportInvalidScopeDisables": undefined,
  "reportNeedlessDisables": undefined,
}
`;

exports[`StylelintRunner should return processed lint results from Stylelint with configured rules 1`] = `
{
  "diagnostics": [
    {
      "code": "block-no-empty",
      "codeDescription": {
        "href": "https://stylelint.io/user-guide/rules/block-no-empty",
      },
      "message": "Unexpected empty block (block-no-empty)",
      "range": {
        "end": {
          "character": 4,
          "line": 0,
        },
        "start": {
          "character": 2,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner should return processed lint results from Stylelint without configured rules 1`] = `
{
  "diagnostics": [
    {
      "code": "CssSyntaxError",
      "message": "Unclosed block (CssSyntaxError)",
      "range": {
        "end": {
          "character": 1,
          "line": 0,
        },
        "start": {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner should throw errors thrown by Stylelint 1`] = `"You must pass stylelint a \`files\` glob or a \`code\` string, though not both"`;
